hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java

```diff
@@ -110,8 +127,110 @@ public SqlResult execute(@Nonnull SqlStatement statement) {
             return res;
         } catch (Exception e) {
             RuntimeException error = rethrow(e, connection);
-            res.onExecuteError(error);
-            throw error;
+            SqlResubmissionResult resubmissionResult = resubmitIfPossible(res, error);
+            if (resubmissionResult == null) {
+                res.onExecuteError(error);
+                throw error;
+            }
+            res.onResubmissionResponse(resubmissionResult);
+            return res;
+        }
+    }
+
+    SqlResubmissionResult resubmitIfPossible(SqlClientResult result, RuntimeException error) {
+        if (!shouldResubmit(error) || !shouldResubmit(result)) {
+            return null;
+        }
+
+        SqlResubmissionResult resubmissionResult = resubmitIfPossible0(result, error);
+        if (resubmissionResult.getSqlError() != null) {
+            SqlError sqlError = resubmissionResult.getSqlError();
+            throw new HazelcastSqlException(
+                    sqlError.getOriginatingMemberId(),
+                    sqlError.getCode(),
+                    sqlError.getMessage(),
+                    null,
+                    sqlError.getSuggestion()
+            );
+        }
+        return resubmissionResult;
+    }
+
+    @SuppressWarnings({"BusyWait", "checkstyle:cyclomaticcomplexity"})
+    private SqlResubmissionResult resubmitIfPossible0(SqlClientResult result, RuntimeException error) {
+        long resubmissionStartTime = System.nanoTime();
+        int invokeCount = 0;
+
+        SqlResubmissionResult resubmissionResult = null;
+        do {
+            ClientConnection connection = null;
+            try {
+                connection = getQueryConnection();
+                QueryId queryId = QueryId.create(connection.getRemoteUuid());
+                logFinest(logger, "Resubmitting query: %s with new query id %s", result.getQueryId(), queryId);
+                result.setQueryId(queryId);
+                ClientMessage message = invoke(result.getSqlExecuteMessage(queryId), connection);
+                resubmissionResult = createResubmissionResult(message, connection);
+                if (resubmissionResult.getSqlError() == null) {
+                    logFinest(logger, "Resubmitting query: %s ended without error", result.getQueryId());
+                } else {
+                    logFinest(logger, "Resubmitting query: %s ended with error", result.getQueryId());
+                }
+                if (resubmissionResult.getSqlError() == null || !shouldResubmit(resubmissionResult.getSqlError())) {
+                    return resubmissionResult;
+                }
+            } catch (Exception e) {
+                logFinest(logger, "Resubmitting query: %s ended with exception", result.getQueryId());
+                RuntimeException rethrown = connection == null ? (RuntimeException) e : rethrow(e, connection);
```

Comment:
are you sure there can be only RuntimeException? In future it may change. Safer would be to use rethrow(e) instead of casting, by default it doesn't wrap if given exception is already a RuntimeException

Reply:
Casting is done when ```connection == null```. The only way that to happen is when: ```connection = getQueryConnection();``` fails on the first iteration. That method can throw only RE:
```java
    public ClientConnection getQueryConnection() {
        try {
            ClientConnection connection = client.getConnectionManager().getConnectionForSql();

            if (connection == null) {
                throw rethrow(QueryException.error(CONNECTION_PROBLEM, "Client is not connected"));
            }

            return connection;
        } catch (Exception e) {
            throw rethrow(e);
        }
    }

    RuntimeException rethrow(Throwable cause) {
        // Make sure that AccessControlException is thrown as a top-level exception
        if (cause.getCause() instanceof AccessControlException) {
            return (AccessControlException) cause.getCause();
        }

        return QueryUtils.toPublicException(cause, getClientId());
    }
```